src/transform/src/projection_pushdown.rs

```diff
@@ -0,0 +1,401 @@
+                let mut actual_projection =
+                    desired_projection.iter().cloned().collect::<HashSet<_>>();
+                actual_projection.extend(0..group_key.len());
+                get_sorted_vec(actual_projection)
+            }
+            MirRelationExpr::TopK {
+                input,
+                group_key,
+                order_key,
+                ..
+            } => {
+                // Group and order keys must be retained, as they define
+                // which rows are retained.
+                let mut columns_to_pushdown =
+                    desired_projection.iter().cloned().collect::<HashSet<_>>();
+                columns_to_pushdown.extend(group_key.iter().cloned());
+                columns_to_pushdown.extend(order_key.iter().map(|o| o.column));
+                // If the `TopK` does not have any new column demand, just push
+                // down the desired projection. Otherwise, push down the sorted
+                // column demand.
+                let columns_to_pushdown = if columns_to_pushdown.len() == desired_projection.len() {
+                    desired_projection.clone()
+                } else {
+                    get_sorted_vec(columns_to_pushdown)
+                };
+                reverse_permute_columns(
+                    group_key
+                        .iter_mut()
+                        .chain(order_key.iter_mut().map(|o| &mut o.column)),
+                    columns_to_pushdown.iter(),
+                );
+                self.action(input, columns_to_pushdown.clone(), gets);
+                columns_to_pushdown
+            }
+            MirRelationExpr::Negate { input } => {
+                self.action(input, desired_projection.clone(), gets);
+                desired_projection.clone()
+            }
+            MirRelationExpr::Union { base, inputs } => {
+                self.action(base, desired_projection.clone(), gets);
+                for input in inputs {
+                    self.action(input, desired_projection.clone(), gets);
+                }
+                desired_projection.clone()
+            }
+            MirRelationExpr::Threshold { input } => {
+                // Threshold requires all columns, as collapsing any distinct values
+                // has the potential to change how it thresholds counts. This could
+                // be improved with reasoning about distinctness or non-negativity.
+                let arity = input.arity();
+                self.action(input, (0..arity).collect(), gets);
+                (0..arity).collect()
+            }
+            MirRelationExpr::ArrangeBy { input, keys } => {
+                // If a key set contains keys that are not demanded upstream,
+                // delete the key set.
+                // If `keys` becomes empty, delete the arrangeby, and keep
+                // pushing the projection. Otherwise, do not push the project past
+                // the ArrangeBy
```

Comment:
This is an interesting opinion that idk what to say about it. We potentially have cases where we want to have an arrangement, even if the specific keys are not important. It could be that this is harmless, of course! But it does change the current definition of `ArrangeBy` which is "ensure these arrangements exists". We might want to document that this is no longer the case if we want to keep this. Does it do important things?

Reply:
This behavior does not do important things. With our current optimizer, it would be a bug if a key set contained keys that are not demanded upstream. 

My rationale for choosing this behavior is that if we do end up with a transform that creates an `ArrangeBy` where the specific keys don't matter, deleting the not-demanded key sets will hopefully signal to this `ArrangeBy`-creating transform that it should choose a new set of keys to arrange on.

Reply:
Well, I like the idea that we make some noise if we have an arrangement that cannot be used. I suspect at some point in the future we'll want to have "any well-distributed arrangement, say by all columns" and we would need to revisit this. Happy either to land this with a big note added to https://github.com/MaterializeInc/materialize/issues/223#issuecomment-904317658, or to not land it.

Reply:
Not landing the arrangeby deletion. I guess we can come back to it if we ever see a Mir with arrangements on unnecessary columns.